src/agent/debugger/src/target.rs

```diff
@@ -272,24 +207,159 @@ impl Target {
+                                    module.new_breakpoint(
+                                        breakpoint.id(),
+                                        breakpoint.kind(),
+                                        sym.address(),
+                                    );
+                                    resolved = true;
+                                }
+                                Err(_) => {
+                                    debug!(
+                                        "unknown symbol {}!{}",
+                                        module.name().display(),
+                                        sym_info.function()
+                                    );
+                                }
+                            }
+                        }
+                        Err(e) => {
+                            error!("Can't set symbolic breakpoints: {}", e);
+                        }
+                    }
+                }
+            }
+            breakpoint::ExtraInfo::Resolved(_) => unreachable!(
+                "address breakpoints should never be added to the unresolved breakpoint list"
+            ),
         }
 
-        Ok(())
+        resolved
+    }
+
+    pub fn new_symbolic_breakpoint(
+        &mut self,
+        id: BreakpointId,
+        module: impl AsRef<str>,
+        func: impl AsRef<str>,
+        kind: BreakpointType,
+    ) -> Result<BreakpointId> {
+        self.maybe_sym_initialize()?;
+        let bp = Breakpoint::from_symbol(id, kind, module.as_ref(), func.as_ref());
+        if !self.try_resolve_unresolved_breakpoint(&bp) {
+            self.unresolved_breakpoints.push(bp);
+        }
+        Ok(id)
+    }
+
+    pub fn new_rva_breakpoint(
+        &mut self,
+        id: BreakpointId,
+        module: impl AsRef<str>,
+        rva: u64,
+        kind: BreakpointType,
+    ) -> BreakpointId {
+        let bp = Breakpoint::from_rva(id, kind, module.as_ref(), rva);
+        if !self.try_resolve_unresolved_breakpoint(&bp) {
+            self.unresolved_breakpoints.push(bp);
+        }
+        id
+    }
+
+    pub fn new_absolute_breakpoint(
+        &mut self,
+        id: BreakpointId,
+        address: u64,
+        kind: BreakpointType,
+    ) -> Result<BreakpointId> {
+        self.module_from_address(address)
+            .new_breakpoint(id, kind, address);
+        Ok(id)
+    }
+
+    fn module_base_from_address(&self, address: u64) -> Result<NonZeroU64> {
+        let dbghelp = dbghelp::lock().expect("can't lock dbghelp to find module");
+        dbghelp.get_module_base(self.process_handle, address)
     }
 
-    pub fn breakpoint_set_at_addr(&self, address: u64) -> bool {
-        self.breakpoints.contains_key(&address)
+    fn module_from_address(&mut self, address: u64) -> &mut Module {
+        let module_base = self
+            .module_base_from_address(address)
+            .unwrap_or(unsafe { NonZeroU64::new_unchecked(module::UNKNOWN_MODULE_BASE_ADDRESS) });
```

Comment:
How about avoiding `unsafe` and doing:
```suggestion
            .unwrap_or(NonZeroU64::new(module::UNKNOWN_MODULE_BASE_ADDRESS).unwrap());
```
Alternately, it looks like `new_unchecked()` is a `const fn`, so maybe we could just make the exported constant a `NonZeroU64`.

Reply:
We have too many `unwrap` calls already and I started replacing them with `expect` or adding proper error handling.

Unfortunately `const fn` doesn't alleviate the need to use `unsafe`, and you can't use `unwrap` or `expect` on a `const`.

Reply:
Ah, darn. Maybe someday soon: https://github.com/rust-lang/rust/issues/51999#issuecomment-818925156